### ISSUE TITLE
LLM pressure test: candidate models on our real workloads (eval harness)

### ISSUE THREAD

**Issue by rodneygagnon** [OpusPopuli/opuspopuli#1142]
## Context

Qwen's viability is in question and open-provenance models are the candidate
replacements. Concrete failure evidence exists (below), but most defects fixed
around the time this was filed were pipeline bugs, not model bugs — so the
decision needs a harness that separates the two, run on our workloads and our
hardware, not benchmark hearsay.

The philosophical pull toward open models is real and specific to this
platform: fully open weights + data + training code is the only model class
where "don't take our word for it" extends to the model itself. The capability
risk was equally real: open models had trailed on structured/JSON output —
exactly what the extraction pipeline depends on.

**A first pass of this harness has now been run (2026-09-10). Results are
below; they retire some assumptions in the original framing and replace the
candidate list.**

## Observed qwen3.5:9b failures to encode as golden cases

1. **Legistar structural manifest** chose the "Upcoming Meetings" widget
   (1–3 rows) over the main `gridCalendar` (16 meeting rows) — with
   self-reported confidence 0.9. Wrong, plausible, self-assured. The exact
   page is the regression fixture; correct answer known.
2. **Bio claims attribution**: "7 claims (2 source, 5 training)" for a county
   supervisor — 5 claims from training memory, not the scraped page. For a
   checkability-first product, training-sourced claims are unverifiable
   claims. Claims-precision is a first-class metric.

## Candidate list (corrected 2026-09-10)

The original list (`qwen3.5:9b`, `qwen3.5:35b`, `olmo2:13b`, `olmo2:32b`) is
stale on both sides — production actually runs **`qwen3.6:35b-a3b`**, and OLMo
2 has been superseded. Current candidates, all Apache 2.0:

| Model | Architecture | Provenance |
|---|---|---|
| `qwen3.6:35b-a3b` | MoE, 3B active | prod baseline; Alibaba |
| `qwen3.5:9b` | dense | dev baseline; Alibaba |
| `olmo-3.1:32b-instruct` | dense | Ai2 — weights + data + training code |
| `olmo-3:7b-instruct` / `olmo-3:7b-think` | dense | Ai2 — same full openness |
| `granite4:32b-a9b-h` | MoE, 9B active | IBM — published data recipe, IP indemnity |
| `gpt-oss:20b` | MoE, 3.6B active | OpenAI, US |

Excluded on licence/openness grounds: Llama (community licence, not OSI),
Gemma (use restrictions). Excluded on provenance grounds alongside qwen:
DeepSeek, GLM, Kimi.

## Harness requirements (learned the hard way — do not skip)

The first run of this harness produced **entirely invalid results** because of
configuration, not model behaviour. Anyone re-running it must:

1. **Set `think` explicitly per model.** The production Ollama provider sends
   `think: false` (`ollama.provider.ts:271`). Omitting it leaves reasoning
   *on* for reasoning models: qwen3.5:9b spent all 2000 tokens on hidden
   reasoning and returned an **empty** `response`, scoring 0/3 on JSON — while
   production generates these analyses fine. `gpt-oss:20b` did the same at
   default effort (22,000–26,000 chars of reasoning, empty answer); it needs
   `think: "low"`. Reasoning variants need a larger budget so thinking cannot
   starve the answer.
2. **Use the real token budget.** `PROPOSITION_ANALYSIS_MAX_TOKENS` defaults
   to **6000**, not 2000.
3. **Use the real prompt.** Templates are published in the public
   `prompt-service` repo (`prisma/seed.ts`) — use them, not the stub
   fallbacks in the local `prompt_templates` table.
4. **Do not score field completeness.** See fabrication finding below.

## Metrics (revised)

Original list stands — JSON-validity rate, field-level correctness, claims
precision, hallucination rate, latency + tok/s on our hardware, context
handling — **plus** these, which the first run showed are necessary:

- **Numeric grounding / fabrication.** Every dollar figure, magnitude and
  percentage the model emits must appear in the source text. This is also a
  prototype of #1209's numeric-consistency validator.
- **Abstention correctness.** An empty `fiscalImpact` on a source containing
  no fiscal analysis is the *right* answer. Scoring field presence alone
  actively rewards fabrication — see below.
- **Claim-span anchoring.** Does `fullText[sourceStart:sourceEnd]` actually
  support the claim?
- **Symmetry / framing / omission / uncertainty calibration** (from the
  unassailability standard) — equivalent questions asked from opposing
  political perspectives must be treated symmetrically. #1140's titler over
  politically opposed measures is the first symmetry case. Not yet measured.
- Adversarial-input fixtures per the #1143 decision (published prompts
  recruit red-teamers; verification gates carry the defence).

## First results — 2026-09-10

MacBook Pro M4 Pro 48GB. Five real propositions from the dev corpus
(2,799–4,477 chars), production `document-analysis-proposition-analysis`
template (6,830 chars), one run each.

| Model | Params (active) | JSON | Anchored | Fabricated | tok/s | s/measure |
|---|---|---|---|---|---|---|
| `granite4:32b-a9b-h` | 32B (9B) MoE | 5/5 | **3/14 · 21%** | 0 | 23.8 | 46 |
| `olmo-3:7b-think` | 7B dense | 4/5 | 2/10 · 20% | 0 | 36.4 | 212 |
| `olmo-3:7b-instruct` | 7B dense | 5/5 | 2/17 · 12% | **1** | 39.2 | **23** |
| `olmo-3.1:32b-instruct` | 32B dense | 5/5 | 1/35 · 3% | 0 | 10.4 | 131 |
| `qwen3.5:9b` | 9B dense | 5/5 | 0/30 · 0% | 0 | 25.7 | 68 |
| `gpt-oss:20b` | 20B (3.6B) MoE | 5/5 | 0/15 · 0% | 0 | **45.3** | 29 |

### The structured-output risk is retired

Every candidate produced valid JSON 5/5 (olmo-3:7b-think 4/5) once configured
correctly. The original concern — that open models would fail the
structured-output requirement the pipeline depends on — did not materialise on
this workload.

### MoE vs dense is the decisive throughput axis

`granite4:32b-a9b-h` and `olmo-3.1:32b-instruct` are **identical in total
parameters (32B) and on-disk size (19.5GB)**. Granite activates 9B per token,
OLMo activates all 32B. Granite is **2.3× faster per token and 2.8× faster per
measure**. Production runs a 3B-active MoE, so a dense-32B migration is a far
steeper regression than raw size suggests — and it lands on the bills sync,
which already takes ~48h at `BILL_ENRICHMENT_CONCURRENCY=1`.

`olmo-3:7b-instruct` is the efficiency surprise: **5.7× faster than
`olmo-3.1:32b-instruct`** (23s vs 131s per measure) with identical Ai2
full-openness provenance.

### Fabrication tracks model size

Two models invented numbers absent from the source:

- `granite4:latest` (3.4B micro, since dropped) — *"$1.2 million in new state
  revenue annually…"* for a measure whose text contains **no `$` at all**.
- `olmo-3:7b-instruct` — *"67%"* in a summary, not present in the source.

Everything ≥9B stayed grounded. Two consequences: a **floor on model size**
for civic analysis, and the grounding check belongs in the **pipeline**, not
only in evals.

### Field completeness is a trap metric

The 3.4B granite topped the first scoreboard at 16/18 fields **because it
fabricated the fiscal impact**, while qwen scored lower for correctly
returning an empty `fiscalImpact` — the honest answer, since AG-filed
initiative text carries no fiscal analysis. Note **all six models** returned
empty `fiscalImpact` on all five measures; that is a source-data property, not
a model failure, and a naive eval would misread it as one.

### Claim-span anchoring is broken across every model — 8/121 (under 7%)

The highest score anywhere is 21%. Failure mode is fabricated offsets, not
near-misses:

- qwen emits neat sequential partitions — `260..580`, `580..850`, `850..1300`
  — i.e. it is *partitioning the document*, not locating text.
- granite cited `1240..5400` and `2700..3600` in a **2,799-character**
  document. Those positions do not exist.

`normalizePayload` only *clamps* offsets into range, so `1240..5400` silently
becomes `1240..2799` — a 1,559-char "citation" spanning half the measure,
rendered to citizens as precise attribution.

**Reasoning does not fix it.** Comparing the two OLMo 7B variants isolates
reasoning cleanly (same family, same size): anchoring rate roughly doubled
(12% → 20%) at **9× the wall-clock** (23s → 212s), with lower JSON reliability
(5/5 → 4/5) and 40% fewer claims. 20% correct citations is still a broken
feature.

This is not a model-selection problem. Tracked as a contract change in
#1212 (quote-then-locate), and it gates #1209.

## Recommendation

**Do not pick a model on this evidence yet.** The harness measures form,
grounding and speed — not analytical quality, which needs the hand-verified
ground truth this issue calls for and which does not exist yet. Directionally:
`gpt-oss:20b` leads on speed, `granite4:32b-a9b-h` on anchoring-plus-balance,
`olmo-3:7b-instruct` on openness-per-second; `olmo-3.1:32b-instruct` — the
model the target architecture nominates — is slowest and near-worst at
anchoring, with provenance as its only remaining argument.

Caveats: five measures, one prompt family, single run each, one machine.
Run-to-run variance is large (qwen anchored 3/16 in one pass and 0/30 in the
next under identical settings), so single-run rankings are directional only.

## Data classification

Eval fixtures are public civic records. No user data enters prompts (existing
posture unchanged).

Labels: infra, knowledge, region

**Comment by rodneygagnon** [comment]
Reframe per the unassailability standard: this eval does not decide WHETHER to move off qwen — model provenance decides that axis on its own (fully-open weights/data/training is the only class where "audit us" extends to the model, and a foreign-trained model on ballot analysis is an attack that works regardless of behavior). The eval measures the CAPABILITY COST of the move and where mitigations must land (larger OLMo tier, task-splitting, stronger verification gates). Corollary: invest the findings into the verification layer — claims-precision gates, fail-closed structural analysis — because unassailability that lives in the pipeline survives every future model change.

**Comment by rodneygagnon** [comment]
**R3 of `docs/plans/ai-architecture-roadmap.md` (approved 2026-09-11)** revises this spec before implementation. Changes from the original issue text:

1. **Candidate set is three, not two, and OLMo 2 is out**: `qwen3.5` (incumbent), **OLMo 3.1 32B Instruct**, **OLMo 3.1 32B Think**. Instruct/Think are separate checkpoints, not a runtime flag — the `think` option in `GenerateOptions` is an Ollama-ism nothing sets, and reasoning-mode choice becomes *which model you load*.
2. **Quantization and runtime are recorded per result, and comparisons are matched-quant.** Ollama serves q4 by default; an OLMo-q4 vs qwen-q4 comparison is partly a quantization comparison. The Studio (128GB) can serve 8-bit via an **MLX / vLLM-Metal eval-only sidecar on a separate port** (`mlx-community/Olmo-3.1-32B-Instruct-8bit` exists; vllm-project/vllm-metal is real, macOS 15+, continuous batching). Production stays on Ollama; the sidecar is instrumentation, not a serving decision. `getModelName()` provenance must distinguish model+quant+runtime or #1149's caches blend incomparable outputs.
3. **The harness gains an embedding calibration set** — the #1074 petition golden set plus an EN/ES query set — so the R5 legacy-embedding convergence has an instrument for recalibrating `MIN_VERIFIED_SIMILARITY` (measured on #1156: nomic scores correct matches at 0.4–0.5, i.e. *at or below* the current 0.50 gate — unrecalibrated cutover sends petition verification dark). One harness serves both M5-residue calibration and M7 selection.
4. **A throughput baseline precedes any runtime discussion**: `OLLAMA_NUM_PARALLEL` is set nowhere and every app-side concurrency knob is 1. Set it to 4, match one worker, re-time a representative job. This sizes M6's "adversarial review doubles inference" constraint with data and decides whether continuous batching arguments are worth anything.
5. Unchanged and reaffirmed: EN+ES parity as a gate, adversarial fixtures (#1143), the unassailability metrics (partisan symmetry via #1140's titler as first case, omission, framing, calibration, source hierarchy), ~50–100 gold items with schema-first growth.

Corpus-quality caveat for gold items: 52/64 proposition summaries are currently title-echo + boilerplate (see the R2 issue) — gold items authored against today's `summary` field inherit that ceiling. Author retrieval gold items against `fullText` spans or post-R2 summaries where possible.

**Comment by rodneygagnon** [comment]
## Handoff: what the eval harness can do as of 2026-09-14

This issue was last updated 2026-09-11. Since then the harness grew an OCR leg
and a set of findings that change R3's own candidate list. All of it currently
lives in PR bodies (#1243, #1245) rather than here, so this is the summary a
fresh context needs.

### Running it

```bash
pnpm --filter @opuspopuli/eval-harness eval:retrieval            # embeddings vs typed queries
pnpm --filter @opuspopuli/eval-harness eval:baseline             # same, xenova/MiniLM
pnpm --filter @opuspopuli/eval-harness eval:ocr -- --engine X    # photo -> OCR -> embed -> retrieve
```

`--engine` accepts `text` (replay stored OCR), `tesseract`, `ollama-vision`,
`olmocr`, `olmocr-mlx`. Results land in `packages/eval-harness/results/*.json`.

### ⚠️ Prerequisite that will look like a bug

**`eval:ocr` throws if prompt-service is unreachable or unseeded:**

```
Prompt template "ocr-transcription" not found in database
```

That is deliberate, not a failure (#1246, #1249). The OCR instructions live in
prompt-service; this repo keeps no inline copy, and the client has no fallback
by design. Set `PROMPT_SERVICE_URL` (+ `PROMPT_SERVICE_API_KEY`) or seed
`prompt_templates` from the prompt-service repo. Every OCR run records
`promptName` / `promptHash` / `promptVersion` in its results JSON, so a score
can be traced to the exact instruction that produced it.

### What the OCR leg measures, and why it is not CER

The gold label is "which measure is this a photograph of", which the
photographer knows — so no hand-transcription is needed. Scoring is top-1 and
margin over the real 64-measure corpus. Character error rate would be expensive
and would measure the wrong thing: the product does not need correct text, it
needs text that RETRIEVES the right measure.

### Results (node, live corpus, production embedding model)

| engine | latency | rank | margin |
|---|---|---|---|
| qwen2.5vl:7b | 16-21s | **1** (every run) | 0.152-0.159 |
| qwen3.6:35b-a3b | 27.7s | 1 | 0.187 |
| Tesseract | ~5s | — | **matched 0 of 5 real production scans** |
| olmOCR-2 (3 GGUF builds) | — | 42-43 | unusable |

Caveats that matter for anyone citing these:

- **qwen2.5vl paraphrases at the margins** — "EXTRACTION" for "ESTABLISHES",
  "address" for "driver's license". Fine for retrieval, NOT fine for anything
  quoted back to a citizen as the filed record.
- **olmOCR-2 is off-distribution on photographs.** Its pipeline expects a PDF
  page plus anchor text from the embedded text layer; a photo has neither. That
  fits the observed bimodality — two clean transcriptions, then deterministic
  garbage at identical parameters.
- **A broken vision GGUF is indistinguishable from a working one** by every
  piece of metadata Ollama exposes. All three olmOCR/Molmo builds advertise
  `vision` through `ollama show`; the only test is sending an image and reading
  the output.
- **`think: false` is mandatory** on reasoning-capable models. With thinking on,
  qwen3.6 spent its entire 1400-token budget reasoning and returned an EMPTY
  response with `done_reason=length` — reads as model incompatibility, is one
  flag.

### This changes R3's candidate set

This issue lists **OLMo 3.1 32B Instruct / Think**. For any vision or OCR
evaluation that is a dead end: **OLMo has no vision capability at all.** Every
working local VLM path measured here is Qwen-derived — olmOCR-2 is itself a
Qwen2.5-VL fine-tune, and Molmo2-8B reports arch `qwen3`. Molmo's OLMo-backed
variant has no usable conversion (`GGML_ASSERT` crash).

OLMo remains the intended main inference model; the provenance boundary shipped
in v1.27.x confines Qwen to OCR via `OCR_VISION_MODEL`, pinned separately from
`LLM_MODEL` precisely so the OLMo switch cannot silently break scanning.

### Remaining R3 work, unchanged

Exit is "≥50 gold items; symmetry metrics produce numbers". Currently **22
items** (14 EN + 8 ES) and **no symmetry metrics** — the partisan-symmetry,
omission, framing, calibration and source-hierarchy measures are all still to
build. #1140's titler is the first symmetry case; adversarial fixtures per
#1143.

Also still open and unbuilt: the Tesseract cross-check as a censorship/omission
guard alongside the VLM (E-24 in the v1.27.1 validation pack). It needs its
divergence metric calibrated in the harness before it is worth wiring up — a
natural fit for whoever picks up the symmetry metrics.

### Not replayable, plan around it

The #1074 petition golden set cannot be reconstructed: scan images are never
persisted by design (#1075). Recalibrating `MIN_VERIFIED_SIMILARITY` needs
petitions re-photographed. The two fixtures in `packages/eval-harness/fixtures`
are a real photograph of a **blank** 25-0007A1 (verified — every signature,
name and address field empty) plus a cropped variant.
